source/utils/bitset.hpp

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#define sizeof_bits(x) (CHAR_BIT * sizeof(x))
+
+#include <cstdint>
+#include <limits>
+#include <stdexcept>
+#include <type_traits>
+#include <concepts>
+#include <spdlog/fmt/fmt.h>
+
+#include "logger/logger.hpp"
+
+namespace sim {
+
+template <typename T>
+concept BitStorageType = 
+    std::is_same_v<T, std::uint8_t>  ||
+    std::is_same_v<T, std::uint16_t> ||
+    std::is_same_v<T, std::uint32_t> ||
+    std::is_same_v<T, std::uint64_t> ||
+    std::is_same_v<T, unsigned __int128>;
+
+template <BitStorageType BitStorage>
+class BitSet {
+public:
+    BitSet() : m_data(0) {};
+    explicit BitSet(BitStorage initial) : m_data(initial) {};
+
+    bool set_bit(std::uint8_t pos, bool value) {
+        if (pos >= sizeof_bits(m_data)) {
+            LOG_ERROR(fmt::format("Bit position is out of range. Max possible position is {}, but got {}", sizeof(BitStorage) - 1, pos));
+            return false;
+        }
+
+        if (value) {
+            m_data |= (1U << pos);
+        } else {
+            m_data &= ~(1U << pos);
+        }
+
+        return true;
+    };
+
+    bool set_range(std::uint8_t low, std::uint8_t high, BitStorage value){
+        if (high >= sizeof_bits(m_data) || low > high) {
+            LOG_ERROR(fmt::format("Range edges error. Low and high should be less than {} and low should be <= hight. Got low = {} and high = {}", sizeof(BitStorage), low, high));
+            return false;
+        }
+
+        const std::uint32_t length = high - low + 1;
+        const std::uint32_t max_val = max_range_value(length);
+
+        if (value > max_val) {
+            LOG_ERROR(fmt::format("Value is out of range. Max possible value = {}, when got {}", max_val, value));
+            return false;
+        }
+
+        std::uint32_t mask = max_val << low;
+        m_data = (m_data & ~mask) | (value << low);
+
+        return true;
+    };
+
+    std::uint8_t get_bit(std::uint8_t pos) const {
```

Comment:
Just a suggestion for better readability:

```suggestion
    book get_bit(std::uint8_t pos) const {
```